quodlibet/quodlibet/util/collection.py

```diff
@@ -595,38 +612,148 @@ def from_songs(cls, dir_, songs, library=None):
+        path = os.path.join(self.dir, self.filename_for(new_name))
         if os.path.exists(path):
             raise ValueError(
-                    _("A playlist named %s already exists.") % new_name)
+                _("A playlist named %(name)s already exists at %(path)s")
+                % {"name": new_name, "path": path})
         return new_name
 
     def delete(self):
         super(FileBackedPlaylist, self).delete()
-        self.__delete_file(self.filename)
+        self._delete_file(self.path)
 
     @classmethod
-    def __delete_file(cls, fn):
+    def _delete_file(cls, fn):
         try:
             os.unlink(fn)
         except EnvironmentError:
             pass
 
     def write(self):
-        fn = self.filename
+        fn = self.path
         with open(fn, "wb") as f:
             for song in self._list:
                 if isinstance(song, str):
                     f.write(fsn2bytes(song, "utf-8") + b"\n")
                 else:
                     f.write(fsn2bytes(song("~filename"), "utf-8") + b"\n")
         if self._last_fn != fn:
-            self.__delete_file(self._last_fn)
+            self._delete_file(self._last_fn)
             self._last_fn = fn
+
+
+class XSPFBackedPlaylist(FileBackedPlaylist):
+    EXT = "xspf"
+    CREATOR_PATTERN = Pattern("<artist|<artist>|<~people>>")
+    _SAFER = {c: quote(c, safe='')
+              for c in ("\\/:*?\"<>|" if is_windows() else "\0/")}
+
+    @classmethod
+    def from_playlist(cls, old_pl: FileBackedPlaylist, library):
+        """Migrate from an existing file-based playlist"""
+
+        def backup_for(path: str) -> str:
+            base = os.path.join(dirname(path), ".backup")
+            if not exists(base):
+                print_d("Creating playlist backup directory %s" % base)
+                os.mkdir(base)
+            return os.path.join(base, basename(path))
+
+        name = old_pl.name
+        new = XSPFBackedPlaylist.new(old_pl.dir, name, library)
+        new.extend(old_pl)
+        new.write()
+        os.rename(old_pl.path, backup_for(old_pl.path))
+        return new
+
+    def _populate_from_file(self):
+        library = self.library
+        try:
+            tree = ET.parse(self.path)
+            # TODO: validate some top-level tag data
+            node = tree.find("title")
+            if self.name != node.text:
+                print_w("Found name \"%s\" instead of \"%s\" for %s"
+                        % (node.text, self.name, self.path))
+            for node in tree.iterfind('.//track'):
+                location = node.find('location')
+                path = location.text.replace('\n', '').replace('\r', '')
+                if path in library:
+                    self._list.append(library[path])
+                elif library and library.masked(path):
+                    self._list.append(path)
+                else:
+                    # TODO: handle missing playlist items (#3105, #729, #3131)
```

Comment:
I wonder how this would be implemented.. The playlist could be loaded, ignore invalid entries and then merge in the changes, right?

Reply:
Who knows - but it all feels possible